packages/grafana-ui/src/components/VizLayout/VizLayout.tsx

```diff
@@ -0,0 +1,91 @@
+import React, { FC, CSSProperties, ComponentType } from 'react';
+import { useMeasure } from 'react-use';
+import CustomScrollbar from '../CustomScrollbar/CustomScrollbar';
+
+export interface Props {
+  width: number;
+  height: number;
+  legend?: React.ReactElement<VizLayoutSlotProps>;
+  children: (width: number, height: number) => React.ReactNode;
+}
+
+export interface VizLayoutType extends FC<Props> {
+  Slot: ComponentType<VizLayoutSlotProps>;
+}
+
+export const VizLayout: VizLayoutType = ({ width, height, legend, children }) => {
+  const containerStyle: CSSProperties = {
+    display: 'flex',
+    width: `${width}px`,
+    height: `${height}px`,
+  };
+
+  if (!legend) {
+    return <div style={containerStyle}>{children(width, height)}</div>;
+  }
+
+  const { position, maxHeight, maxWidth } = legend.props;
+  const [legendRef, legendMeasure] = useMeasure();
+  let size: VizSize | null = null;
+
+  const vizStyle: CSSProperties = {
+    flexGrow: 1,
+  };
+
+  const legendStyle: CSSProperties = {
+    flexGrow: 1,
+  };
+
+  let positionWithAuto = position;
+  if (position === 'auto') {
+    positionWithAuto = width > height ? 'right' : 'bottom';
+  }
+
+  switch (positionWithAuto) {
+    case 'bottom':
+      containerStyle.flexDirection = 'column';
+      legendStyle.maxHeight = maxHeight;
+
+      if (legendMeasure) {
+        size = { width, height: height - legendMeasure.height };
+      }
+      break;
+    case 'right':
+      containerStyle.flexDirection = 'row';
+      legendStyle.maxWidth = maxWidth;
+
+      if (legendMeasure) {
+        size = { width: width - legendMeasure.width, height };
+      }
+      break;
+  }
+
+  return (
+    <div style={containerStyle}>
+      <div style={vizStyle}>{size && children(size.width, size.height)}</div>
+      {legend && (
+        <div style={legendStyle} ref={legendRef}>
+          <CustomScrollbar>{legend}</CustomScrollbar>
+        </div>
+      )}
+    </div>
```

Comment:
Think this is fine as long as we assume visualization can utilize only a single additional slot (i.e. legend only).

Reply:
With the approach taken in https://github.com/grafana/grafana/blob/master/public/app/plugins/panel/graph3/VizLayout.tsx the idea was that the visualization can have multiple additional eleents in different slots.

Reply:
yea, that was what I can see we need now in new NG Chart & Pie chart. But  would be nice to have the API extend to more slots in the future without API changes

Reply:
Ah, and one more thing I forgot with my approach. It allows you to render Graph plugins as children of a graph component, as the actual layou is rendered as such:
```
 <GraphNG data={data.series} timeRange={timeRange} timeZone={timeZone} {...canvasSize}>
            {builder.addSlot('canvas', <Canvas />).render()}
```

But I guess we can **maybe** get rid of this as the Legend does not have to be rendered as uPlot plugin as it unly uses `usePlotData` hook which is not necessary there. It can just accept the data frame pushed to the GraphPanel instead.

Reply:
Ok, I understand. 

But yea, if avoiding complexity like that sounds like a good thing.

I really do not understand how that even works though with uPlot and new graph panel, the legend has to render first in order to know the canvas size. But it if the legend is a uplot plugin that renders as part of the uplot then uplot needs to render first with some initial size and then again after the legend size has been determined?

Reply:
it probably makes sense to have a distinction between a uPlot plugin and GraphNG/UPlotChart plugin.

uPlot needs an explicit width/height in pixels, so anything you need to fit next to it must be known in advance.

since you guys render the legend (and probably title) externally, `Legend` would be a `GraphNG` plugin that ends up rendering first and then passing the remainder of the container size to init uPlot afterwards.

i'm assuming there are no plans to actually allow authoring and passing raw uPlot plugins through this component stack, but just create purpose-build GraphNG plugins with well-defined APIs?

Reply:
> since you guys render the legend (and probably title) externally, Legend would be a GraphNG plugin that ends up rendering first and then passing the remainder of the container size to uPlot afterwards.

Yea exactly, Grafana needs to render the legend first to get the available size remaining for the chart/viz

Reply:
probably all slots that must be implicitly sized. this might present some challenges if you plan to use nested `VizLayout`s that contain other graphs, but that seems pretty edge case.

in this case grid-template-areas may come to the rescue in terms of layout not governing component order, since you can render components in implict -> explicit order and accumulate their dimensions before rendering the center, though i'm not sure if they will be measurable until the next event loop tick :/

1. top slot
2. right slot
3. bottom slot
4. left slot
5. center slot (explicit, pass accumulated dims)